File: src/main/java/com/example/dpop/orchestrator/session/ProcessSession.java
package com.example.dpop.orchestrator.session;

import jakarta.persistence.Column;
import jakarta.persistence.DiscriminatorColumn;
import jakarta.persistence.DiscriminatorType;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.Lob;
import jakarta.persistence.Table;
import jakarta.persistence.Version;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "process_session")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "purpose", discriminatorType = DiscriminatorType.STRING)
public abstract class ProcessSession {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "process_session_id", nullable = false)
    private UUID processSessionId;

    @Column(name = "channel_session_id", nullable = false)
    private UUID channelSessionId;

    @Enumerated(EnumType.STRING)
    @Column(name = "purpose", nullable = false, length = 20, insertable = false, updatable = false)
    private ProcessPurpose purpose;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private ProcessStatus status;

    @Column(name = "account_id")
    private Long accountId;

    @Column(name = "selected_identification_method", length = 50)
    private String selectedIdentificationMethod;

    @Column(name = "selected_authentication_method", length = 50)
    private String selectedAuthenticationMethod;

    @Column(name = "person_id")
    private Long personId;

    @Lob
    @Column(name = "pending_challenge")
    private String pendingChallenge;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "expires_at", nullable = false)
    private Instant expiresAt;

    @Column(name = "consumed_at")
    private Instant consumedAt;

    @Version
    @Column(name = "version", nullable = false)
    private Long version;

    protected ProcessSession() {
    }

    protected ProcessSession(UUID channelSessionId, ProcessPurpose purpose, Instant expiresAt) {
        this.processSessionId = UUID.randomUUID();
        this.channelSessionId = channelSessionId;
        this.purpose = purpose;
        this.status = ProcessStatus.ACTIVE;
        this.createdAt = Instant.now();
        this.expiresAt = expiresAt;
    }

    public UUID getProcessSessionId() {
        return processSessionId;
    }

    public UUID getChannelSessionId() {
        return channelSessionId;
    }

    public ProcessPurpose getPurpose() {
        return purpose;
    }

    public ProcessStatus getStatus() {
        return status;
    }

    public void setStatus(ProcessStatus status) {
        this.status = status;
    }

    public Long getAccountId() {
        return accountId;
    }

    public void setAccountId(Long accountId) {
        this.accountId = accountId;
    }

    public String getSelectedIdentificationMethod() {
        return selectedIdentificationMethod;
    }

    public void setSelectedIdentificationMethod(String selectedIdentificationMethod) {
        this.selectedIdentificationMethod = selectedIdentificationMethod;
    }

    public String getSelectedAuthenticationMethod() {
        return selectedAuthenticationMethod;
    }

    public void setSelectedAuthenticationMethod(String selectedAuthenticationMethod) {
        this.selectedAuthenticationMethod = selectedAuthenticationMethod;
    }

    public Long getPersonId() {
        return personId;
    }

    public void setPersonId(Long personId) {
        this.personId = personId;
    }

    public String getPendingChallenge() {
        return pendingChallenge;
    }

    public void setPendingChallenge(String pendingChallenge) {
        this.pendingChallenge = pendingChallenge;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getExpiresAt() {
        return expiresAt;
    }

    public Instant getConsumedAt() {
        return consumedAt;
    }

    public void consume() {
        this.consumedAt = Instant.now();
        this.status = ProcessStatus.COMPLETED;
    }

    public Long getVersion() {
        return version;
    }

    public boolean isExpired() {
        return Instant.now().isAfter(expiresAt);
    }

    public boolean isConsumed() {
        return consumedAt != null;
    }
}

File: src/main/kotlin/com/example/dpop/orchestrator/session/ProcessSession.kt
package com.example.dpop.orchestrator.session

import jakarta.persistence.Column
import jakarta.persistence.DiscriminatorColumn
import jakarta.persistence.DiscriminatorType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Inheritance
import jakarta.persistence.InheritanceType
import jakarta.persistence.Table
import jakarta.persistence.Version
import java.time.Instant
import java.util.UUID

/**
 * Abstract base with shared lifecycle, account reference and routing state
 * (docs/02-domaenenmodell.md #2). Process-specific fields live only on the
 * concrete subclasses, never here.
 */
@Entity
@Table(name = "process_session")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "purpose", discriminatorType = DiscriminatorType.STRING)
abstract class ProcessSession protected constructor(
    @Column(name = "channel_session_id", nullable = false)
    var channelSessionId: UUID? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "purpose", nullable = false, length = 20, insertable = false, updatable = false)
    var purpose: ProcessPurpose? = null,

    @Column(name = "expires_at", nullable = false)
    var expiresAt: Instant? = null
) {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "process_session_id", nullable = false)
    var processSessionId: UUID? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "state", nullable = false, length = 20)
    var state: ProcessState? = null

    @Column(name = "account_id")
    var accountId: Long? = null

    // Routing state (docs/04-orchestrierung.md): where the client goes next.
    @Column(name = "next_type", length = 10)
    var nextType: String? = null

    @Column(name = "next_tool_id", length = 50)
    var nextToolId: String? = null

    /**
     * Which ToolSession is actually authorized to act as [nextToolId] right now - not just any
     * ToolSession row with a matching toolId. Without this, an orphaned/superseded ToolSession
     * (e.g. from a duplicate activation request) would keep passing a toolId-only check even
     * though its own tool-module data was never populated, surfacing as a confusing
     * "Unknown tool session" error deep inside the module instead of a clean 409 here.
     */
    @Column(name = "next_tool_session_id")
    var nextToolSessionId: UUID? = null

    @Column(name = "next_context", length = 50)
    var nextContext: String? = null

    @Column(name = "next_step", length = 50)
    var nextStep: String? = null

    @Column(name = "created_at", nullable = false)
    var createdAt: Instant? = null

    @Column(name = "consumed_at")
    var consumedAt: Instant? = null

    @Version
    @Column(name = "version", nullable = false)
    var version: Long? = null

    init {
        state = ProcessState.STARTED
        createdAt = Instant.now()
    }

    /**
     * [toolSessionId] is null for a fresh offer (the tool hasn't been activated yet - a client
     * must still call activate to mint one) and non-null when an already-active ToolSession is
     * merely progressing to its next internal step (docs/06-ablaeufe.md).
     */
    fun setNextTool(toolId: String, step: String, toolSessionId: UUID? = null) {
        nextType = "tool"
        nextToolId = toolId
        nextToolSessionId = toolSessionId
        nextContext = null
        nextStep = step
    }

    fun setNextFlow(context: String, step: String) {
        nextType = "flow"
        nextToolId = null
        nextToolSessionId = null
        nextContext = context
        nextStep = step
    }

    fun consume() {
        consumedAt = Instant.now()
        state = ProcessState.CONSUMED
    }

    fun fail() {
        state = ProcessState.FAILED
    }

    /** User-initiated abandonment, distinct from Failed (retry-exhausted) - docs/02-domaenenmodell.md #4. */
    fun cancel() {
        state = ProcessState.CANCELLED
    }

    val isExpired: Boolean
        get() = expiresAt?.let { Instant.now().isAfter(it) } ?: false

    val isConsumed: Boolean
        get() = consumedAt != null
}
